site/content/api/Tenant-API.md

```diff
@@ -0,0 +1,264 @@
+
+| Name        | Mandatory | Location                 | Type     | Description |
+| :---------- | :-------: | :----------------------- | :------- | :---------- |
+| *subject*   | yes       | *properties*             | *string* | MUST be set to `get`. |
+
+The body of the message SHOULD be empty and will be ignored if it is not.
+
+**Response Message Format**
+
+A response to a *get tenant information* request contains the [Standard Response Properties]({{< relref "#standard-response-properties" >}}).
+
+The response message includes payload as defined in the [Tenant Format]({{< relref "#tenant-format" >}}) section below.
+
+The response message's *status* property may contain the following codes:
+
+| Code | Description |
+| :--- | :---------- |
+| *200* | OK, the payload contains the tenant information for the requested tenant. |
+| *404* | Not Found, there is no tenant identified with the given *tenant-id*. |
+
+For status codes indicating an error (codes in the `400 - 499` range) the message body MAY contain a detailed description of the error that occurred.
+
+## Update Tenant
+
+Clients use this command to *update* information about an already existing tenant. All of the information that has been previously provided for the tenant gets *replaced* with the information contained in the request message.
+
+This operation is *optional*, implementors of this API may provide other means for updating tenant information, e.g. a RESTful API or a configuration file.
+
+**Message Flow**
+
+The following sequence diagram illustrates the flow of messages involved in a *Client* updating tenant information.
+
+![Update Tenant Information message flow](../tenant_UpdateTenantSuccess.png)
+
+**Request Message Format**
+
+The following table provides an overview of the properties a client needs to set on an *update tenant* message in addition to the [Standard Request Properties]({{< relref "#standard-request-properties" >}}).
+
+| Name        | Mandatory | Location                 | Type     | Description |
+| :---------- | :-------: | :----------------------- | :------- | :---------- |
+| *subject*   | yes       | *properties*             | *string* | MUST be set to `update`. |
+
+The request message MUST include payload as defined in the [Tenant Format]({{< relref "#tenant-format" >}}) section below.
+ 
+**Response Message Format**
+
+A response to an *update tenant* request contains the [Standard Response Properties]({{< relref "#standard-response-properties" >}}).
+
+The response message's *status* property may contain the following codes:
+
+| Code  | Description |
+| :---- | :---------- |
+| *204* | No Content, the tenant has been updated successfully. |
+| *404* | Not Found, there is no tenant identified with the given *tenant-id*. |
+
+For status codes indicating an error (codes in the `400 - 499` range) the message body MAY contain a detailed description of the error that occurred.
+
+## Remove Tenant
+
```

Comment:
I have concerns here:
Removal of tenants sounds quite simple, but looking at the overall system behaviour it is not IMHO.
What happens with open connections from protocol adapters when the tenant is removed?
Possibly a tenant should only be marked as deleted, but not removed.

On the one hand this is "only" an API that handles more or less the CRUD operations of tenants, but on the other hand it should give at least hints on obvious problems that can occur in a detailed implementation (finally leaving open how the implementation is done).
Can you please add some lines for that?

Reply:
Marking the tenant as *deleted*. however, is an implementation detail that is left up to the implementer. I do agree with @sysexcontrol, though, that we should outline the semantics and potential impact of *removing* a tenant.

Reply:
That's a good point :-) I guess we have some similar effects for the remove operation already in the Device and Credentials API too. (When credentials are removed and a device still has a valid JWT token I assume it still will be able to send data) 

I assume in the hono system we have some sort of eventual consistency model.
As the tenant api  is not actively pushing updates to other services the tenant will in time disappear in the system. How fast this happen depends on the implementation. 

I added some lines to make it more clear what the effects of the removal are.